GSoC 2023: JSR-305 probe blog

Jagrutiti opened the pull request:
Submitting blog for `JSR-305` probe as a part of GSoC activity.

Jagrutiti commented:
Hi everyone,

For most of the probes that I work on, I only have one or two lines of information as to why they are important. I do not really have the background information as to why the probes were suggested or how could they prove the most beneficial. 

That is why the blogs aren't really that long. My main goal is to document the working and thought processes that went behind building them. 

Thanks.

krisstern commented:
@aneveux You may want to give this a review when you have the time

zbynek commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,39 @@
+
+== Introduction
+
+`@Nonnull` and `@CheckForNull` annotations (part of JSR-305) have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016]. For security reasons it is important that the outdated annotations are highlighted and replaced with `javax.annotation.Nonnull` and `edu.umd.cs.findbugs.annotations.NonNull`
```
"For security reasons" -- that sounds misleading. The real reasons are mentioned in https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ.
Also `javax.annotation.Nonnull` is one of the outdated annotations, nothing should be replaced with it.

Jagrutiti replied:
I have added more information. Thanks for reviewing.

aneveux commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,52 @@
+
+`@Nonnull` and `@CheckForNull` annotations (part of JSR-305) have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].
+
+The `javax.annotation.Nonnull` import should be replaced with `edu.umd.cs.findbugs.annotations.NonNull`.
```
Why giving this example and not the replacement for `CheckForNull`? Not sure it is very important to have here?

kmartens27 commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,51 @@
+
+This probe will identify the outdated annotations used in the plugin and highlight them in the scoring system.
+
+== Challenges and Learning
+
+The outdated annotations are imported in `.java` files. My first approach was to go for an abstract class because I assumed there could be other probes that may require to look for code in `.java` files.
+After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule] all imports from the `javax.annotation` package should be checked for deprecation.
+
+Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It") which means develop features only when you need it, the `abstract` wasn't the current requirement. I restructured the code and used a single class to implement the probe.
```
```suggestion
== Challenges and Learning

The outdated annotations are imported in `.java` files.
My first approach was to go for an abstract class because I assumed there could be other probes that may need to look for code in `.java` files.
After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule], all imports from the `javax.annotation` package should be checked for deprecation.

Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It"), which means features are developed only when you need them, the `abstract` wasn't the current requirement.
I restructured the code and used a single class to implement the probe.
```

suggestion for formatting and some syntax

Were there any additional challenges in restructuring the code or using a single class?

Jagrutiti replied:
There weren't any other challenges. I am just sharing the thought process of arriving at the solution here to help the future GSoCers/contributors.

kmartens27 commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,51 @@
+
+Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It") which means develop features only when you need it, the `abstract` wasn't the current requirement. I restructured the code and used a single class to implement the probe.
+
+== Conclusion and Outcome
+This probe will identify the deprecated annotations and let the maintainers know which imports in a plugin require replacement.
```
```suggestion
== Conclusion and Outcome
This probe will identify the deprecated annotations and let maintainers know which imports in a plugin require replacement.
```

Suggestion for syntax

kmartens27 reviewed (commented):
thanks very much @Jagrutiti, I provided some feedback on a few syntax/formatting instances, and was curious if there was any additional information/context you could provide in regards to the issues/challenges faced and the importance of the probe.

Please feel free to share any questions or concerns, and I'm happy to help in any way!

zbynek reviewed (approved):
The issues I mentioned are now resolved, thanks. It might make sense to move the context of the article (link to your GSoC project page) from the last paragraph to the first one so that readers know what is the article about.

Jagrutiti commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,51 @@
+
+== Introduction
+
+The `JSR-305` framework provides annotations to detect software defects. Its executive committee voted it as dormant since May 2012 (see: link:https://jcp.org/en/jsr/detail?id=305[JSR 305: Annotations for Software Defect Detection
+])
+
```
In the above link: https://jcp.org/en/jsr/detail?id=305 the reason is mentioned:

> Reason: The Executive Committee voted to list this JSR as dormant in May 2012

Jagrutiti commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,51 @@
+
+== Importance of the probe
+
+This probe will identify the outdated annotations used in the plugin and highlight them in the scoring system.
```
The scoring system hasn't been implemented yet. I do not have more information about it.

Jagrutiti commented:
> from the last paragraph to the first one so that readers know what is the article about.

Where do you suggest I add in the first paragraph? I am introducing the probe there.

zbynek commented:
@Jagrutiti the first paragraph could start with something like "This blog post describes the latest addition to the plugin health scoring system developed as part of link:...[my Google summer of Code Project]: a probe that detects usages of the JSR-305 framework in plugins."

Jagrutiti commented:
I updated the project doc links for other blogs as well. As they were incorrect. 

It was a minor update I changed them in this PR. I hope it is okay.

kmartens27 commented on content/blog/2023/07/16/2023-07-16-security-scan-probe-blog.adoc:
```diff
@@ -50,4 +50,4 @@ During this project, I accomplished the following:
 == Links
 
-For more information or if you have any questions, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to  "Plugin Health Score"] description page.
+For more information or if you have any questions, please visit the official GSoC 2023 project link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[Adding Probes to  "Plugin Health Score"] description page.
```
```suggestion
For more information or if you have any questions, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to  "Plugin Health Score"] description page.
```
The modified link does not direct anywhere, is there a reason why you modified the links this way? The way it is written originally works and goes to the project page as intended.

Jagrutiti replied:
The links weren't working in my local with the old syntax. That is why I changed it.

I just checked the published blogs and it works fine. 

I think I am missing something here.

zbynek replied:
Links generally should *not* work in GitHub, you either have to run the site locally with `make run` or check on the preview page https://deploy-preview-6653--jenkins-io-site-pr.netlify.app/blog/2023/09/03/jsr305-probe-blog/ (note that URLs on the site do not end with `.adoc`)

Jagrutiti replied:
IntelliJ IDE renders the `.adoc` syntax so I expect it to render even the URLs. That's where I was wrong.

kmartens27 commented on content/blog/2023/07/16/2023-07-16-unreleased-production-changes-probe.adoc:
```diff
@@ -79,5 +79,5 @@ Starting from ground zero is tough, whether implementing a new Java library or s
 
 == Links
-For more information or to find answers to any questions you might have, head to the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to  "Plugin Health Score"] description page.
+For more information or to find answers to any questions you might have, head to the official GSoC 2023 project link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[Adding Probes to  "Plugin Health Score"] description page.
```
```suggestion
For more information or to find answers to any questions you might have, head to the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to  "Plugin Health Score"] description page.
```
Same suggestion as the other link, it appears to work as originally written so it's not clear why the change was made.

kmartens27 commented on content/blog/2023/07/20/2023-07-20-third-party-repository-detection-probe.adoc:
```diff
@@ -61,4 +61,4 @@ By doing so, we can easily list the third-party repositories used in every pom f
 
-For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
\ No newline at end of file
+For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[Adding Probes to "Plugin Health Score"] description page.
```
```suggestion
For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
```
Same as other link suggestions

kmartens27 commented on content/blog/2023/08/28/2023-08-28-number-of-open-issues-probe.adoc:
```diff
@@ -51,4 +51,4 @@ This will also keep the plugin relevant to the latest Jenkins updates.
 
-For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
\ No newline at end of file
+For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[Adding Probes to "Plugin Health Score"] description page.
```
```suggestion
For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
```
Same as other link suggestions

kmartens27 commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,56 @@
+
+== Introduction
+
+This blog post describes the latest addition to the plugin health scoring system developed as part of link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[my Google summer of Code Project]: a probe that detects usages of the `JSR-305` framework in plugins.
```
```suggestion
This blog post describes the latest addition to the plugin health scoring system developed as part of link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[my Google summer of Code Project]: a probe that detects usages of the `JSR-305` framework in plugins.
```
Same as other link suggestions

kmartens27 commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,56 @@
+
+For more insight, refer to the link:https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ[Jenkins developer mailing list conversation].
+
+The `@Nonnull` and `@CheckForNull` annotations, which are part of JSR-305, have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].
```
```suggestion
The `@Nonnull` and `@CheckForNull` annotations, which are part of JSR-305, have been link:/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].
```
Suggestion to remove the extraneous part of this link

kmartens27 commented on content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc:
```diff
@@ -0,0 +1,56 @@
+
+== Links
+
+For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[Adding Probes to "Plugin Health Score"] description page.
```
```suggestion
For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
```
Same as other link suggestions

kmartens27 reviewed (commented):
Hi @Jagrutiti, I reviewed the link changes, and I'm not sure what the intent was on those.

When checking the deploy site, it leads to a broken link and this occurs in all the blog posts that were modified. Is the original not leading to the correct page?

krisstern commented:
Looks good to me @Jagrutiti! However, would you like to update the publication date to today before I go ahead and merge it?

Jagrutiti commented:
Hey @krisstern 

Could you confirm if I made the changes you had asked for?

krisstern commented:
Yes, that's correct @Jagrutiti!

krisstern reviewed (approved):
LGTM
